scheduling/allocator.rb

```diff
@@ -163,6 +181,38 @@ def self.candidate_hosts(request)
           .select_append { count.function.*.as(vm_provisioning_count) }
           .where(display_state: "creating"))
 
+      ds = if request.use_slices && request.require_shared_slice
+        # We are looking for hosts that have at least once slice already allocated but with enough room
+        # for our new VM. This means it has to be a sharable slice, with cpu and memory available
+        # We then combine it with search for a host, as usual, with just open space on the host where
+        # we could allocate a new slice
+        # Later in VmHostSliceAllocator the selected hosts will be scored depending if a slice is reused or
+        # new one is created
+        ds.with(:slice_utilization, DB[:vm_host_slice]
+          .select_group(:vm_host_id)
+          .select_append { (sum(Sequel[:total_cpu_percent]) - sum(Sequel[:used_cpu_percent])).as(slice_cpu_available) }
+          .select_append { (sum(Sequel[:total_memory_gib]) - sum(Sequel[:used_memory_gib])).as(slice_memory_available) }
+          .where(enabled: true)
+          .where(is_shared: true)
+          .where(Sequel[:used_cpu_percent] + request.cpu_percent_limit <= Sequel[:total_cpu_percent])
+          .where(Sequel[:used_memory_gib] + request.memory_gib <= Sequel[:total_memory_gib]))
+          # end of 'with'
```

Comment:
```suggestion
            .select_group(:vm_host_id)
            .select_append { (sum(Sequel[:total_cpu_percent]) - sum(Sequel[:used_cpu_percent])).as(slice_cpu_available) }
            .select_append { (sum(Sequel[:total_memory_gib]) - sum(Sequel[:used_memory_gib])).as(slice_memory_available) }
            .where(enabled: true)
            .where(is_shared: true)
            .where(Sequel[:used_cpu_percent] + request.cpu_percent_limit <= Sequel[:total_cpu_percent])
            .where(Sequel[:used_memory_gib] + request.memory_gib <= Sequel[:total_memory_gib])
          )
```

Seems better to use indentation to signal the end of `with`.

Reply:
I indented the lines, but the editor (Rubycop?) forces the closing parenthesis to be on the last where. I cannot make it go to the new line. So I left the comment in there as well, but hopefully it looks a bit better.